### PR TITLE
Add "set" methods to ProtectedAsMock

### LINKED ISSUE #1164: Missing setter functionality from ProtectedAsMock

### PR DESCRIPTION
**Pull request by tonyhallett**
Fix #1164

### CODE REVIEW

File: tests/Moq.Tests/ProtectedAsMockFixture.cs [hunk 1]
```diff
@@ -213,6 +213,89 @@ public void SetupSequence_can_setup_actions()
...
+				o => o[
+					It.IsInRange(0, 5, Range.Inclusive),
+					It.IsIn("Bad", "JustAsBad")
+				] = It.Is<int>(i => i > 10)
+			).Throws(ExpectedException.Instance);
+
+			mock.Object.SetMultipleIndexer(1, "Ok", 999);
+
+			Assert.Throws<ExpectedException>(() => mock.Object.SetMultipleIndexer(1, "Bad", 999));
+
```
Comment (stakx) [inline tests/Moq.Tests/ProtectedAsMockFixture.cs:290]: Detail: for consistency with existing formatting, please remove this superfluous blank line.

File: tests/Moq.Tests/ProtectedAsMockFixture.cs [hunk 2]
```diff
@@ -213,6 +213,89 @@ public void SetupSequence_can_setup_actions()
...
+			VerifySet(); // AtLeastOnce
+			
+			mock.Object.SetMultipleIndexer(1, "JustAsBad", 12);
+			VerifySet(Times.Exactly(2));
+
+			Assert.Throws<MockException>(() => VerifySet(Times.AtMostOnce()));
+
+			var mockException = Assert.Throws<MockException>(() => VerifySet(Times.AtMostOnce(),"custom fail message"));
+			Assert.StartsWith("custom fail message", mockException.Message);
+
```
Comment (stakx) [inline tests/Moq.Tests/ProtectedAsMockFixture.cs]: Detail: for consistency with existing formatting, please remove this superfluous blank line.

File: tests/Moq.Tests/ProtectedAsMockFixture.cs [hunk 3]
```diff
@@ -298,6 +381,10 @@ public void VerifyGet_includes_failure_message_in_exception()
 			Assert.Contains("Was not queried.", exception.Message);
 		}
 
+		public interface INested
+		{
+			int Value { get; set; }
+		}
 		public abstract class Foo
```
Comment (stakx) [inline tests/Moq.Tests/ProtectedAsMockFixture.cs]: Detail: for consistency with existing formatting, please add a blank line between those two lines of code.

**Review by stakx — CHANGES_REQUESTED** [review]
Apologies for keeping you waiting so long. I've been rather busy at work.

Good work! :+1: I'd be happy if you could make a few small changes (as per the review comments). Also, could you please add an entry to `CHANGELOG.md` in the _Unreleased_ section, for instance:

```diff
 ## Unreleased

+#### Added
+
+* `SetupSet`, `VerifySet` methods for `mock.Protected().As<>()` (@tonyhallett, #1165)

 #### Fixed
```

Thank you!

**Comment by tonyhallett** [comment]
@stakx 
Changes made.  There is now one duck replacer class that now correctly accounts for virtual protected accessors and automocking #1162 

Once this has been approved I will update the draft InSequence for protected mocks pull request #1174

File: CHANGELOG.md [hunk 4]
```diff
@@ -7,8 +7,13 @@ The format is loosely based on [Keep a Changelog](http://keepachangelog.com/en/1
 
 ## Unreleased
 
+#### Added
+
+* `SetupSet`, `VerifySet` methods for `mock.Protected().As<>()` (@tonyhallett, #1165)
+
 #### Fixed
 
+* Virtual properties and automocking not working for `mock.Protected().As<>()` (@tonyhallett, #1165)
```
Comment (stakx) [inline CHANGELOG.md]: I'm a little confused... what is this doing here all of a sudden? How is it related to the addition of `SetupSet` and `VerifySet`?

I'd prefer if we kept PRs focused on one thing at a time &mdash; this makes PRs much easier to understand and review.

Can you please remove all changes associated with this seemingly unrelated bug fix from this PR, then re-submit it as a separate PR? (I've marked the suspected changes with a &dagger; in the other review comments.)

File: tests/Moq.Tests/ProtectedAsMockFixture.cs [hunk 5]
```diff
@@ -130,6 +130,13 @@ public void Setup_can_setup_generic_method()
 			Assert.Equal(3, handledMessages.Count);
 		}
 
+		[Fact]
+		public void Setup_can_automock()
+		{
+			this.protectedMock.Setup(m => m.Nested.Method(1)).Returns(123);
+			Assert.Equal(123, mock.Object.GetNested().Method(1));
+		}
+
```
Comment (stakx) [inline tests/Moq.Tests/ProtectedAsMockFixture.cs]: &dagger;) Is this part of the bugfix mentioned in the changelog? If so, please re-submit as part of a separate bugfix PR.

File: src/Moq/Protected/ProtectedAsMock.cs [hunk 6]
```diff
@@ -239,11 +246,21 @@ protected override Expression VisitMethodCall(MethodCallExpression node)
 				{
 					var targetParameter = Expression.Parameter(this.targetType, left.Name);
 					return Expression.Call(targetParameter, FindCorrespondingMethod(node.Method), node.Arguments);
-				}
+				}
```
Comment (stakx) [inline src/Moq/Protected/ProtectedAsMock.cs]: Detail: please revert this whitespace-only change.

File: src/Moq/Protected/ProtectedAsMock.cs [hunk 7]
```diff
@@ -305,7 +322,7 @@ private PropertyInfo FindCorrespondingProperty(PropertyInfo duckProperty)
 			{
 				var candidateTargetProperties =
 				    this.targetType
-				    .GetProperties(BindingFlags.NonPublic | BindingFlags.Instance)
+				    .GetProperties(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance)
```
Comment (stakx) [inline src/Moq/Protected/ProtectedAsMock.cs]: I'm not sure I understand this change.

* How is it related to the addition of `SetupSet`, `VerifySet`?
* My memory may be a bit rusty, but this looks incorrect to me: we're looking for a `protected` member, so why do we need `BindingFlags.Public`?

File: tests/Moq.Tests/ProtectedAsMockFixture.cs [hunk 8]
```diff
@@ -150,6 +157,26 @@ public void SetupGet_can_setup_readwrite_property()
 			Assert.Equal(42, actual);
 		}
 
+		[Fact]
+		public void SetUpGet_can_automock()
```
Comment (stakx) [inline tests/Moq.Tests/ProtectedAsMockFixture.cs]: Detail: `Setup`, not `SetUp`.

File: tests/Moq.Tests/ProtectedAsMockFixture.cs [hunk 9]
```diff
@@ -150,6 +157,26 @@ public void SetupGet_can_setup_readwrite_property()
...
+		[Fact]
+		public void SetupGet_can_setup_virtual_property()
+		{
+			this.protectedMock.SetupGet(m => m.VirtualGet).Returns(42);
+
+			var actual = mock.Object.GetVirtual();
+
+			Assert.Equal(42, actual);
+		}
+
```
Comment (stakx) [inline tests/Moq.Tests/ProtectedAsMockFixture.cs]: &dagger;) Is this part of the bugfix mentioned in the changelog? If so, please re-submit as part of a separate bugfix PR.

File: tests/Moq.Tests/ProtectedAsMockFixture.cs [hunk 10]
```diff
@@ -384,12 +418,51 @@ public void VerifyGet_includes_failure_message_in_exception()
 		public interface INested
 		{
 			int Value { get; set; }
+			int Method(int value);
```
Comment (stakx) [inline tests/Moq.Tests/ProtectedAsMockFixture.cs:421]: &dagger;) Is this part of the bugfix mentioned in the changelog? If so, please re-submit as part of a separate bugfix PR.

File: src/Moq/Protected/ProtectedAsMock.cs [hunk 11]
```diff
@@ -255,7 +272,7 @@ protected override Expression VisitMember(MemberExpression node)
 				}
 				else
 				{
-					return node;
+					return base.VisitMember(node);
```
Comment (stakx) [inline src/Moq/Protected/ProtectedAsMock.cs]: &dagger;) Is this part of the bugfix mentioned in the changelog? If so, please re-submit as part of a separate bugfix PR.

File: tests/Moq.Tests/ProtectedAsMockFixture.cs [hunk 12]
```diff
@@ -384,12 +418,51 @@ public void VerifyGet_includes_failure_message_in_exception()
...
+				{
+					_virtualGet = value;
+				}
+
+			}
+
+			public int GetVirtual()
+			{
+				return VirtualGet;
+			}
```
Comment (stakx) [inline tests/Moq.Tests/ProtectedAsMockFixture.cs]: &dagger;) Is this part of the bugfix mentioned in the changelog? If so, please re-submit as part of a separate bugfix PR.

File: tests/Moq.Tests/ProtectedAsMockFixture.cs [hunk 13]
```diff
@@ -437,10 +510,17 @@ public void SetMultipleIndexer(int index, string sIndex, int value)
...
+
+			public int GetMultipleIndexer(int index, string sIndex)
+			{
+				return this[index, sIndex];
+			}
 		}
 
 		public interface Fooish
 		{
+			int VirtualGet { get; set; }
```
Comment (stakx) [inline tests/Moq.Tests/ProtectedAsMockFixture.cs]: †) Is this part of the bugfix mentioned in the changelog? If so, please re-submit as part of a separate bugfix PR.

File: tests/Moq.Tests/ProtectedAsMockFixture.cs [hunk 14]
```diff
@@ -437,10 +510,17 @@ public void SetMultipleIndexer(int index, string sIndex, int value)
 			{
 				this[index, sIndex] = value;
 			}
+
+			public int GetMultipleIndexer(int index, string sIndex)
+			{
+				return this[index, sIndex];
+			}
```
Comment (stakx) [inline tests/Moq.Tests/ProtectedAsMockFixture.cs]: This doesn't appear to be related with the addition of `SetupSet` and `VerifyGet`, and in either case, it doesn't appear to get used anywhere. Please remove.

File: src/Moq/Protected/ProtectedAsMock.cs [hunk 15]
```diff
@@ -239,11 +246,21 @@ protected override Expression VisitMethodCall(MethodCallExpression node)
 				{
 					var targetParameter = Expression.Parameter(this.targetType, left.Name);
 					return Expression.Call(targetParameter, FindCorrespondingMethod(node.Method), node.Arguments);
-				}
+				} 
 				else
 				{
-					return node;
+					return base.VisitMethodCall(node);
```
Comment (stakx) [inline src/Moq/Protected/ProtectedAsMock.cs]: †) Is this part of the bugfix mentioned in the changelog? If so, please re-submit as part of a separate bugfix PR.

**Review by stakx — CHANGES_REQUESTED** [review]
Hi @tonyhallett, thanks for making the requested changes!

I'm a little confused by the additional changes you've made. Can we please do the bugfixing of auto-mocking and virtual properties in one or more separate focused PRs... or is this somehow a necessary consequence of mixing your replacer class into the existing `DuckReplacer`?

File: src/Moq/Protected/ProtectedAsMock.cs (hunk 7 above)
Comment (tonyhallett) [reply src/Moq/Protected/ProtectedAsMock.cs]: > How is it related to the addition of SetupSet, VerifySet?

This is the fix for virtual property with protected accessor and applies equally for SetupSet as it does for SetupGet.

> My memory may be a bit rusty, but this looks incorrect to me: we're looking for a protected member, so why do we need BindingFlags.Public?

Yes we are looking for a protected accessor of a public property.

File: CHANGELOG.md (hunk 4 above)
Comment (tonyhallett) [reply CHANGELOG.md]: > or is this somehow a necessary consequence of mixing your replacer class into the existing DuckReplacer?

Indeed it is.  The automocking and virtual properties issues apply equally here.  

> Can you please remove all changes associated with this seemingly unrelated bug fix from this PR, then re-submit it as a separate PR?

They are related but to make it clear that there are bugs that apply to the current functionality - https://github.com/moq/moq4/issues/1162 then I will make a separate pull request and when that has been merged in I will proceed with this pull request.

Comment (tonyhallett) [reply CHANGELOG.md]: https://github.com/moq/moq4/pull/1185

File: src/Moq/Protected/ProtectedAsMock.cs (hunk 7 above)
Comment (tonyhallett) [reply src/Moq/Protected/ProtectedAsMock.cs]: This BindingFlags change makes ProtectedAsMock behave the same as ProtectedMock

https://github.com/moq/moq4/blob/a6fde8b6d79a7437bf642d115785b97f40779b6a/src/Moq/Protected/ProtectedMock.cs#L346

```
		private static PropertyInfo GetProperty(string propertyName)
		{
			return typeof(T).GetProperty(
				propertyName,
				BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public);
		}
```

File: CHANGELOG.md (hunk 4 above)
Comment (stakx) [reply CHANGELOG.md:16]: @tonyhallett, thanks for the confirmation, and for your patience! I've merged your other PR (great work!), so you can go ahead cleaning up this one. I'll try my best to respond more quickly this time!

File: src/Moq/Protected/ProtectedAsMock.cs (hunk 7 above)
Comment (stakx) [reply src/Moq/Protected/ProtectedAsMock.cs]: Without reviewing the existing code I couldn't say in which places `ProtectedAsMock` needs to be the same as `ProtectedMock` and where it may legitimately diverge, but your first explanation (part of your bug fix) is good enough for me! :-)

**Comment by stakx** [comment]
@tonyhallett, thanks for extracting the bug fix into a separate PR #1185. (Good work!) I've marked those review comments that were related to it as resolved, so you can now proceed with rebasing & cleaning up this PR; there are now only a few details remaining. I'll try my best to respond more quickly this time!

File: tests/Moq.Tests/ProtectedAsMockFixture.cs (hunk 14 above)
Comment (tonyhallett) [reply tests/Moq.Tests/ProtectedAsMockFixture.cs]: done

File: src/Moq/Protected/ProtectedAsMock.cs (hunk 6 above)
Comment (tonyhallett) [reply src/Moq/Protected/ProtectedAsMock.cs]: done

**Comment by tonyhallett** [comment]
@stakx You ok if I squash these commits into one ?

**Comment by stakx** [comment]
@tonyhallett, sure, go ahead. Let me know once it's ready to be merged.

**Review by stakx — APPROVED** [review]

**Comment by tonyhallett** [comment]
@stakx 

> You ok if I squash these commits into one ?

Unfortunately this has highlighted my poor knowledge of git.  I assume I need --rebase-merges....
Please can you assist ?

**Comment by stakx** [comment]
@tonyhallett, if you want, I can do the squashing when I merge (GitHub's web UI has an option to squash-merge a PR). That would probably be easiest and quickest.

If you want to do it yourself, for learning purposes, you can do so using an interactive rebase (`git rebase -i`):

```shell
git checkout protectedasmock-set-methods
git rebase -i main  # assuming that you forked `protectedasmock-set-methods` off the `main` branch
# in the editor that opens, change all except the first `pick` command to `squash`
# then save and exit the editor
```

If you want to ensure that your branch is current with respect to this repository (`moq/moq4`)'s `main` branch, you could additionally do the following first:

```shell
git remote add upstream https://github.com/moq/moq4.git
git checkout main
git pull upstream/main
```

This should fast-forward your `main` branch so it's in sync with `moq/moq4`'s `main` branch. Then follow it up with the interactive rebase described above.

**Comment by tonyhallett** [comment]
> if you want, I can do the squashing when I merge (GitHub's web UI has an option to squash-merge a PR). That would probably be easiest and quickest.

Agreed.  Please do

**Comment by stakx** [comment]
Done. 🚀

Thanks for your contribution!
